encoding/ssz/detect/configfork.go

```diff
@@ -0,0 +1,178 @@
+// then Unmarshal()s the type and returns an instance of state.BeaconState if successful.
+func (cf *VersionedUnmarshaler) UnmarshalBeaconState(marshaled []byte) (s state.BeaconState, err error) {
+	forkName := version.String(cf.Fork)
+	switch fork := cf.Fork; fork {
+	case version.Phase0:
+		st := &ethpb.BeaconState{}
+		err = st.UnmarshalSSZ(marshaled)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to unmarshal state, detected fork=%s", forkName)
+		}
+		s, err = v1.InitializeFromProtoUnsafe(st)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to init state trie from state, detected fork=%s", forkName)
+		}
+	case version.Altair:
+		st := &ethpb.BeaconStateAltair{}
+		err = st.UnmarshalSSZ(marshaled)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to unmarshal state, detected fork=%s", forkName)
+		}
+		s, err = v2.InitializeFromProtoUnsafe(st)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to init state trie from state, detected fork=%s", forkName)
+		}
+	case version.Bellatrix:
+		st := &ethpb.BeaconStateBellatrix{}
+		err = st.UnmarshalSSZ(marshaled)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to unmarshal state, detected fork=%s", forkName)
+		}
+		s, err = v3.InitializeFromProtoUnsafe(st)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to init state trie from state, detected fork=%s", forkName)
+		}
+	default:
+		return nil, fmt.Errorf("unable to initialize BeaconState for fork version=%s", forkName)
+	}
+
+	return s, nil
+}
+
+var beaconBlockSlot = fieldSpec{
+	// ssz variable length offset (not to be confused with the fieldSpec offest) is a uint32
+	// variable length. Offsets come before fixed length data, so that's 4 bytes at the beginning
+	// then signature is 96 bytes, 4+96 = 100
+	offset: 100,
+	t:      typeUint64,
+}
+
+func slotFromBlock(marshaled []byte) (types.Slot, error) {
+	slot, err := beaconBlockSlot.uint64(marshaled)
+	if err != nil {
+		return 0, err
+	}
+	return types.Slot(slot), nil
+}
+
+var errBlockForkMismatch = errors.New("fork or config detected from state is different than block")
+
+// UnmarshalBeaconBlock uses internal knowledge in the VersionedUnmarshaler to pick the right concrete SignedBeaconBlock type,
+// then Unmarshal()s the type and returns an instance of block.SignedBeaconBlock if successful.
+func (cf *VersionedUnmarshaler) UnmarshalBeaconBlock(marshaled []byte) (block.SignedBeaconBlock, error) {
+	slot, err := slotFromBlock(marshaled)
+	if err != nil {
+		return nil, err
+	}
+
+	// heuristic to make sure block is from the same version as the state
+	// based on the fork schedule for the Config detected from the state
+	// get the version that corresponds to the epoch the block is from according to the fork choice schedule
+	// and make sure that the version is the same one that was pulled from the state
```

Comment:
This comment mentions a state, but the code does not deal with any states.

Reply:
updated